pkg/token/cache.go

```diff
@@ -38,19 +42,43 @@ func (k CacheKey) String() string {
 	return fmt.Sprintf("{%s:role.%s,%s,%d,%d}", k.Domain, k.Role, k.ProxyForPrincipal, k.MinExpiry, k.MaxExpiry)
 }
 
+func (k CacheKey) Size() uint {
+	structSize := uint(unsafe.Sizeof(k))
+	// unsafe.Sizeof() ONLY count the string struct, need to count the actual string block explicitly
+	stringSize := len(k.Domain) + len(k.ProxyForPrincipal) + len(k.Role)
+	return structSize + uint(stringSize)
+}
+
 type LockedTokenCache struct {
 	cache map[CacheKey]Token
 	lock  sync.RWMutex
+
+	// memoryUsage is the estimated number of bytes used by the cache.
+	memoryUsage int64
 }
 
 func NewLockedTokenCache() *LockedTokenCache {
-	return &LockedTokenCache{cache: make(map[CacheKey]Token)}
+	return &LockedTokenCache{
+		cache:       make(map[CacheKey]Token),
+		memoryUsage: 0,
+	}
 }
 
 func (c *LockedTokenCache) Store(k CacheKey, t Token) {
 	c.lock.Lock()
 	defer c.lock.Unlock()
+	oldToken, ok := c.cache[k]
 	c.cache[k] = t
+
+	// update cache memory usage
+	tokenSize := t.Size()
+	if ok {
+		oldTokenSize := oldToken.Size()
+		c.memoryUsage += int64(tokenSize) - int64(oldTokenSize)
+	} else {
+		keySize := k.Size()
+		c.memoryUsage += int64(keySize + tokenSize)
```

Comment:
```suggestion
		c.memoryUsage += int64(k.Size() + tokenSize)
```

Reply:
or you may change the argument name as `key` not `k`

Reply:
fixed: 655fb92, ddcb83f

Reply:
a bit more here:
https://github.com/AthenZ/k8s-athenz-sia/pull/69#discussion_r1437356525